python/mujinwebstackclient/controllerwebclientraw.py

```diff
@@ -280,3 +349,124 @@ def CallGraphAPI(self, query, variables=None, headers=None, timeout=5.0):
             raise ControllerGraphClientException(_('Unexpected server response %d: %s') % (statusCode, raw), statusCode=statusCode, response=response)
 
         return content['data']
+
+    async def _OpenWebSocketConnection(self):
+        authorization = self._session.auth._GetAuthorization()
+
+        # URL to http GraphQL endpoint on Mujin controller
+        graphEndpoint = '%s/api/v2/graphql' % self._baseurl
+        parsedUrl = urlparse(graphEndpoint)
+        # parse url and handle different scheme
+        webSocketScheme = ''
+        if parsedUrl.scheme == "https":
+            webSocketScheme = 'wss'
+        elif parsedUrl.scheme == "http":
+            webSocketScheme = 'ws'
+        uri = '%s://%s%s' % (webSocketScheme, parsedUrl.netloc, parsedUrl.path)
+
+        adapter = self._session.adapters.get("http://")
+        if isinstance(adapter, UnixSocketAdapter):
+            self._websocket = await websockets.unix_connect(
+                path=adapter.get_unix_endpoint(),
+                uri=uri,
+                subprotocols=['graphql-ws'],
+                additional_headers={
+                    'Content-Type': 'application/json',
+                    'Accept': 'application/json',
+                    'X-CSRFToken': 'token',
+                    'Authorization': authorization,
+                },
+            )
+        else:
+            self._websocket = await websockets.connect(
+                uri=uri,
+                subprotocols=['graphql-ws'],
+                additional_headers={
+                    'Content-Type': 'application/json',
+                    'Accept': 'application/json',
+                    'X-CSRFToken': 'token',
+                    'Authorization': authorization,
+                },
+            )
```

Comment:
You can further simplify this part

```suggestion
        # prepare the headers
        headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
            'X-CSRFToken': 'token',
            'Authorization': authorization,
        }
        subprotocols = ['graphql-ws']

        # decide on using unix socket or not
        adapter = self._session.get_adapter(uri)
        if isinstance(adapter, UnixSocketAdapter):
            self._websocket = await websockets.unix_connect(
                path=adapter.get_unix_endpoint(),
                uri=uri,
                subprotocols=subprotocols,
                additional_headers=headers,
            )
        else:
            self._websocket = await websockets.connect(
                uri=uri,
                subprotocols=subprotocols,
                additional_headers=headers,
            )
```